channel.py

```diff
@@ -377,6 +377,53 @@ def _payload(self) -> JsonType:
         return {"data": (b64encode(json_minify(payload).encode("utf8"))).decode("utf8")}
 
     async def send_watch(self) -> bool:
+        #-------------------------------------------------
+        # FIX 2024/5
+        #  |
+        #  v
+        try:
+            response: JsonType = await self._twitch.gql_request(        # Gets signature and value
+                GQL_OPERATIONS["PlaybackAccessToken"].with_variables({"login": self._login})
+                )
+        except MinerException as exc:
+            raise MinerException(f"Channel: {self._login}") from exc
+        signature: JsonType | None = response["data"]['streamPlaybackAccessToken']["signature"]
+        value: JsonType | None = response["data"]['streamPlaybackAccessToken']["value"]
+        if not signature or not value:
+            return None
+
+        RequestBroadcastQualitiesURL = f"https://usher.ttvnw.net/api/channel/hls/{self._login}.m3u8?sig={signature}&token={value}"
+
+        try:
+            async with self._twitch.request(                            # Gets list of streams
+                "GET", RequestBroadcastQualitiesURL
+            ) as response1:
+                BroadcastQualities = await response1.text()
+        except RequestException:
+            return False
+        
+        LowestQualityBroadcastURL = BroadcastQualities.split("\n")[-1]  # Just takes the last line, this should probably be handled better in the future
+
+        try:
+            async with self._twitch.request(                            # Gets actual streams
+                "GET", LowestQualityBroadcastURL
+            ) as response2:
+                StreamURLList = await response2.text()
+        except RequestException:
+            return False
+
+        StreamLowestQualityURL = StreamURLList.split("\n")[-2] # For whatever reason this includes a blank line at the end, this should probably be handled better in the future
+
+        try:
+            async with self._twitch.request(                            # Downloads the stream
+                "GET", StreamLowestQualityURL
+            ) as response3:                                             # I lied, well idk, but this code doesn't listen for the actual video data
+                return response3.status == 200
+        except RequestException:
+            return False
```

Comment:
I'm not entirely sure what happens in this case. The miner is supposed to be lightweight on the network traffic data it uses. Doing 3 requests every minute doesn't exactly sound lightweight. I can investigate and rewrite this in the future if needed, but repeating data has to be cached where possible.

Reply:
Me neither. The 3 requests are unavoidable, as the signature from the first changes every time, the second one returns a .m3u8 containing the link to the 3rd, which is the current part of the livestream.

As for the last request, as I understand it, it is the request itself that triggers the backend to give watch time (it doesn't progress without it) and I have never heard of sending a request along with something to indicate you don't actually want that request, but I haven't done *that* much with APIs in general, so idk.

Reply:
I seem to get progress replacing "GET" with "HEAD". This can only be done for the last request tho, as we need to actually get the data from the other ones.

Reply:
Measuring time from beginning to end of the watch loop (without the last await) and send_watch(self).
So, at least with my good internet and cpu, the new change is not adding much.
```
requests duration: 1.0315825939178467
watch loop duration: 19.580376863479614
requests duration: 0.9307780265808105
watch loop duration: 10.93083930015564
requests duration: 0.8689451217651367
watch loop duration: 10.870950698852539
requests duration: 0.8665504455566406
watch loop duration: 10.88144564628601
requests duration: 0.894366979598999
watch loop duration: 10.891481161117554
requests duration: 0.8790795803070068
watch loop duration: 10.88401746749878
requests duration: 0.9005200862884521
watch loop duration: 10.91445255279541
```

With request time at 15s (I have that for testing right now), Resource Monitor shows a max Total below 4000B/s for python.exe 
![image](https://github.com/DevilXD/TwitchDropsMiner/assets/72623832/2a7c3b09-0fa0-4b50-823d-cfa10a7a4471)

Reply:
> The miner is supposed to be lightweight on the network traffic data it uses. Doing 3 requests every minute doesn't exactly sound lightweight.

Do we know how much the browser sends when a user is _actually_ watching the stream? - Because it'll be difficult/suspicious to be more "lightweight" than that.

Reply:
> > The miner is supposed to be lightweight on the network traffic data it uses. Doing 3 requests every minute doesn't exactly sound lightweight.
> 
> Do we know how much the browser sends when a user is _actually_ watching the stream? - Because it'll be difficult/suspicious to be more "lightweight" than that.

Well, it has always been way less, meaning it was always "suspicious". If you don't want that, you can make a browser extension, that automatically navigates the twitch site, because otherwise you'd have to essentially write an entire browser into the code to not be "suspicious".

Efficiency is kinda the point, tho this being written in pretty much the slowest mainstream programming language doesn't help.

Reply:
Closing this, as the remaining potential optimization is already tracked here: https://github.com/DevilXD/TwitchDropsMiner/pull/466/files#r1600840668